### PR TITLE
Using setState to render the view

### PR DESCRIPTION
**Pull request by zaaack**
Fix https://github.com/fable-elmish/elmish/issues/60

This issue has been fixed in react long long  ago, but it appears in elmish-react now. If we using setState to update the input's value, it won't reset the cursor, because the real input's value after typing is the same as react vdom's, so react won't change the real dom, and cursor won't change. But if we update the view by `ReactDOM.render`, this issue occurred again, so I think we could change to the commonly used way (update the view by setState) in react to keep the consistency.

Warning: this is a breaking change, in the original way, `DefaultValue`  can work around it just like `Value` in this new way, if you change `DefaultValue` in model, it would update the view, too. But now, `DefaultValue` is the `default value`, and won't sync to view if the view is not first render. If you need to control the input, you should using `Value` instead, which is the same as in normal react app.

### CODE REVIEW

**Comment by MangelMaxime** [comment]
@et1975 For info, 
this is breaking HMR support for Elmish.

And I can't use it on a simple input, it's failing at runtime with ` f is not a function` when I try to dispatch a message into my application. I tried to dispatch a message from a button same problem.

**Comment by zaaack** [comment]
@MangelMaxime Can you try again? Looks I ran into some fable issue maybe, it works now in my elmish-react-template app. Didn't notice HMR is breaking, I'll see what I can do.

**Comment by zaaack** [comment]
Looks like the previous code triggered a wired webpack bug, it translate a function to a IIFE


code from repl:
```js
export const hasUpdate = defaultArg(defaultArg(lastModel, null, function (a) {
  return !(model === a);
} /*a function */), true);
``

code from webpack:
```js
 var hasUpdate = Object(__WEBPACK_IMPORTED_MODULE_8__nuget_packages_fable_core_1_3_0_fable_core_Option__["b" /* defaultArg */])(Object(__WEBPACK_IMPORTED_MODULE_8__nuget_packages_fable_core_1_3_0_fable_core_Option__["b" /* defaultArg */])(lastModel, null, function (a) {
                    return !(model === a);
                }())/* a IIFE */, true);
```

Still trying to understand how HMR works.

**Comment by zaaack** [comment]
@MangelMaxime Now HMR is working, again!

File: src/react.fs [hunk 1]
```diff
@@ -3,24 +3,54 @@ namespace Elmish.React
...
+    }
+
+    type App(props) as this =
+        inherit Component<AppProps, AppState>(props)
+
+        do this.setInitState { view = Unchecked.defaultof<ReactElement> }
+        do (props.getInternalSetState this.setState)
 
+        member this.render () =
+            printfn "render"
```
Comment (MangelMaxime) [inline src/react.fs]: Please remove this debug log :)

**Comment by MangelMaxime** [comment]
HMR seems to works, and I don't see performance problem from my tests.

Will let Eugene take the relay from now.

**Comment by zaaack** [comment]
@MangelMaxime Updated!

**Comment by et1975** [comment]
>This issue has been fixed in react long long ago

Clearly some optimization is taking place, but apparently in React impl it happens at model->VDOM stage, instead of VDOM->DOM (as Preact appears to be doing it). If that's the optimization you want why not implement that specifically - a function that takes some state (`string`) and returns `Prop` or `Prop list` suitable given the difference?

The only reason Native impl uses an app component is because Native doesn't work without. It's ugly and it was a massive pain to get it to work properly. Looking at your code I have no confidence that it will work correctly and you are not inspiring any by letting others find the issues instead of doing thorough testing yourself. Specifically, I find the use of rendered `view` as state troublesome conceptually, I also have a feeling that it would negate the benefits of `lazy` views (`com<>` will always return a new reference). What is the ultimate cost of this? I don't know, it's possible that certain elements (like the ones using D3, which are known to depend on `lazy` will be severely affected).

In general, not that I don't appreciate the effort, but the issue you are fixing is closed and the discussion around it clearly amounts to "there are known workarounds" not "send the PR!".

My suggestion is to keep this as a separate implementation of React renderer, in your own library.

**Comment by zaaack** [comment]
@et1975 Thanks for your patient explanation. What do you mean ` a function that takes some state (string) and returns Prop or Prop list suitable given the difference?` ? I don't think there is a way to change `Value`'s default behavior without modify the library self. 

I know maybe this looks a little hacky, but it totally followed the React's core API 😉. And In my imagine it should works like a drop-in replacement for original implementation, the root component (App) only rendered by ReactDOM.render once, so I don't need to wrap it with lazy view,  I only wrapped it's children with lazyView. React don't do model diff in default, so there should be no cost in putting a big view into state, since it will just update the view after calling `setState`.

This also won't break React's `shouldComponentUpdate` API, it still compose the components like a normal react app, so I think lazyView will still work, and I can confirm it:



> In general, not that I don't appreciate the effort, but the issue you are fixing is closed and the discussion around it clearly amounts to "there are known workarounds" not "send the PR!".

To be clear, there are two main concerns of this PR:

1. People will always run into this issue, especially those comes from React. Then they need to use the `DefaultValue` workaround. `DefaultValue` isn't `value` but work just the same as `value` in normal React, and there even isn't an equivalent to `defaultValue` in React, maybe it's a bit confusing? What's more, what if people are using some third-party UI libraries that doesn't implement `DefaultValue` at all? 

2. Actually it's my first time that knows `ReactDOM.render` would using React's vdom diff, I think most react app only use it as `mount reactApp to some dom node`, so there is still a risk that there were or will be (with React's upgrade) other differences between `ReactDOM.render` and `setState`, and at that time it still need to change back to keep consistency and benefit from react's upgrade. And this is what I worried most.

But I totally agree that this need more feedback and testing, especially elmish-react have a *big* ecosystem and this is a also a breaking change. Maybe release in alpha or beta channel? I'll definitely try and test it with my current fable project.

**Comment by zaaack** [comment]
@alfonsogarciacaro Publishing a fork version is a choice, but I don't have confidence I can get enough feedback from a small community for a fork repo. Currently `DefaultValue` work around is still not a big problem, this pr might more like a prevention of potential issues/risks and user experience enhancement.

I thought React's `setState` is [already well optimized and batched](https://reactjs.org/docs/react-component.html#setstate), so I removed the `requestAnimationFrame code`. But after some digging,  looks adding RAF still can optimize react's performance, and react team are [also considering adding it to react core](https://github.com/facebook/react/issues/11171#issuecomment-335431662). 

Your observable approach is much clear 👍, either of them getting merged is OK for me. Even if neither of these can, I think I can still reference a fork repo by paket, which others can also help testing.

**Comment by et1975** [comment]
For completeness, this is the function that I think replicates the React optimization in its entirety:
```fsharp
    let inline valueOrDefault value e = 
        if e |> isNull |> not && !!e?value <> !!value then e?value <- !!value
```

Usage:
```fsharp
   R.input [Ref (valueOrDefault model.Value); OnChange ... ]
```

**Comment by cgravill** [comment]
Just to add to my previous thumbs up, I'm using your code @et1975 and it works in all use cases I've tried so that's great thanks.

**Pull request closed without merging**